Optimize SkyShader.js with pre-calculated consts

erichlof opened the pull request:
See discussion in #9894

erichlof committed:
Made redundant calculations into consts

erichlof committed:
added const vec3 totalRayleigh and other optimizations

erichlof committed:
comments and cleanup

mrdoob commented:
@brianchirls looks good?

brianchirls commented:
At first glance this looks okay, but I want check some things on IE/Edge and do a more thorough review. Give me a day or so to take a closer look.

erichlof commented:
@brianchirls 
Thanks Brian for taking a closer look.  I used Google's searchbar calculator to input all of the calculations for the equation at https://github.com/erichlof/three.js/blob/dev/examples/js/SkyShader.js#L54-L55  .  I came up with similar values to the "simplifiedRayleigh" so I think I did everything right.  But feel free to check my math - that's one heck of an equation!   ;-)

p.s. there were a couple of minor spots where the shader was calculating pi*2 or 180/pi, but these happened less frequently (maybe even once), so I didn't bother making those consts also.  Again, feel free to change those if you like.

brianchirls commented on examples/js/SkyShader.js:
```diff
@@ -47,17 +47,18 @@ THREE.ShaderLib[ 'sky' ] = {
-		// wavelength of used primaries, according to preetham
-		"const vec3 lambda = vec3(680E-9, 550E-9, 450E-9);",
+		"const vec3 K = vec3(0.686, 0.678, 0.666);", 
+                "const vec3 MieConst = pi * pow((2.0 * pi) / lambda, vec3(v - 2.0)) * K;",
```
This line should be indented with tabs, not spaces

brianchirls commented on examples/js/SkyShader.js:
```diff
@@ -73,7 +74,7 @@ THREE.ShaderLib[ 'sky' ] = {
 		"vec3 totalMie(vec3 lambda, float T)",
```
if `lambda` is no longer used in this function, it should be removed as a parameter.

brianchirls commented:
There are a bunch of lines that are indented by spaces, when they should be tabs...

brianchirls commented on examples/js/SkyShader.js:
```diff
@@ -47,17 +47,18 @@ THREE.ShaderLib[ 'sky' ] = {
+		"const vec3 lambda = vec3(680E-9, 550E-9, 450E-9);",
+		
+                // this pre-calcuation replaces the older totalRayleigh(vec3 lambda) function which was:
+		// return (8.0 * pow(pi, 3.0) * pow(pow(n, 2.0) - 1.0, 2.0) * (6.0 + 3.0 * pn)) / (3.0 * N * pow(lambda, vec3(4.0)) * (6.0 - 7.0 * pn));
```
These constants don't quite match what used to be `simplifiedRayleigh`. What are `n` and `pn` in the function you used to calculate your constants?

brianchirls commented:
I've left a few comments inline. Please reply and I'll do another review.

Also, I don't know if @mrdoob has a style policy for shaders, but there should probably be spaces between arithmetic operators and variables/literals.

erichlof commented:
Hi @brianchirls 
Ok I'll take a look and try to clean up the file along with conforming to MrDoob's code style indentations.  If I recall, n was the refraction index of air or something like that.  I'll dig them up.

erichlof commented:
Found it:
const float pn = 0.035;	// depolatization factor for standard air

I have no idea what that is, but here's the original shader where all this started:
https://github.com/SimonWallner/kocmoc-demo/blob/RTVIS/media/shaders/scattering.glsl

I'm not sure why the constants don't exactly match up with SimplifiedRayleigh.  It's probably due to floating point precision and arturitu's way of calculating the SimplifiedRayleigh:
"vec3 totalRayleigh()",
  "{",		
 	"return 0.0005 / vec3(94, 40, 18);",
  "}",
  The constants that I calculated (by just plugging the original Preetham equation into the Google search bar) are very close though.  Even changing a more significant number doesn't have any visual impact.  The only reason to use my constants is that there are no calculations to do every pixel and every frame of animation.

erichlof committed:
Update SkyShader.js

brianchirls commented:
I hate to nitpick, but I still see some spaces that should be tabs.

Also, I put together this code to calculate `totalRayleigh` based on those constants in the code you referenced:
```javascript
(function rayleigh() {

	const pi = 3.141592653589793238462643383279502884197169;
	const n = 1.0003;
	const pn = 0.035;
	const N = 2.545E25;

	const pow = Math.pow;

	return [680E-9, 550E-9, 450E-9].map(lambda => {
		return (8 * pow(pi, 3) * pow(pow(n, 2) - 1, 2) * (6 + 3 * pn)) / (3 * N * pow(lambda, 4) * (6 - 7 * pn));
	});
}());
// [0.000005804542996261093,
  0.000013562911419845635,
  0.000030265902468824876]
```
It's probably best to write them in "scientific notation" to save space and to be more clear

erichlof commented:
Ok I'll change the constants to scientific notation (that does look a bit nicer!) :-)
Whoops I must have missed the spaces/tabs issue.  It's hard to see in my editor.  Working on it....

brianchirls commented:
Great. Also, please double-check the math, since my numbers are different from yours.

erichlof committed:
scientific notation and tabs rather than spaces

erichlof commented:
I think I fixed all the tabs now.  I converted your results to scientific notation.  I'll double check the calculations.  The Red channel seems close enough, but there are larger differences like you said in the Green and Blue channels.  Again, these numbers are so very small, my answers vs. your answers will have no visual impact that I can tell on my machine.  Still, for correctness-sake, I'll double check...

brianchirls commented:
Hang on. It's not working in MS Edge or MSIE. The old version works, so it must be something in these changes.

@erichlof Do you have a PC you can test/debug with? I'm not seeing any console messages.

erichlof commented:
Hi Brian, could you take a look at the original equation from the Preetham paper?
https://www.cs.utah.edu/~shirley/papers/sunsky/sunsky.pdf
It's in the middle of page 23,  2nd equation.

Notice how
6+3pn 
--------
6-7pn

is grouped in parentheses.  I can't remember my High school multiplication commutative properties, but would that affect the results, since in your code equation, the 6+3pn is multiplied with the terms before it first.  In the paper, it seems the above term is done first by itself, since it's in parentheses.

erichlof commented:
I have IE on Windows 10.  I'll take a look...

brianchirls commented:
The equation checks out. It looks different because of the parentheses, but the results will be the same.

Okay, I found the problem with the Microsoft browsers: they don't like computed constants. So starting with `MieConst`, they need to be entered as numbers. You might be able to do some basic multiplication or addition, but the `pow` is likely a problem. Should be the same in IE as Edge, I think.

So here's what I'd do. First, since you have the paper, it'd be great to post a link to it in the comments. Just let's keep any comments outside the shader string so they can be minified out.

Next, I'd hard-code the values for each of those constants, but I'd place a comment with the original formula above each one. Alternatively, you could compute them in Javascript and place the result in the shader string as a constant value.

erichlof commented:
No, it's not working on IE.  No messages in console.  I'll check older versions of SkyShader to make sure it's something in the PR vs. an already existing problem

erichlof commented:
Oh ok, we were cross posting.  Yes, I'll put all the constants in - I understand what you are saying about IE and Edge and no calculation code in constants.  I'm running out of time for this afternoon unfortunately, I will get back to it tonight.  Thanks!

erichlof committed:
remove calculations inside 'const' declarations

erichlof commented:
It is now working across all browsers on my Windows 10 machine.  I commented out the original calculations and placed them above their respective constant declarations.  Just need verification that the shader works on other platforms and we should be good to go!  :-)

brianchirls commented:
Runs okay on my Mac and my S6. Chrome, FF, Safari. Also fine on my iOS simulator.

erichlof commented:
Yay!  Does everything look ok so far as the PR is concerned?

mrdoob commented:
Thanks!

erichlof commented:
Thanks @mrdoob for merging!  Also many thanks to @brianchirls  for reviewing and helping with this PR, including his solutions for working correctly on IE and Edge!  :)
